library/cc/key_value_store.cc

```diff
@@ -0,0 +1,46 @@
+#include "library/cc/key_value_store.h"
+
+#include "library/common/data/utility.h"
+
+namespace Envoy {
+namespace Platform {
+
+namespace {
+
+envoy_data c_kv_store_read(envoy_data key, const void* context) {
+  auto kv_store = *static_cast<KeyValueStoreSharedPtr*>(const_cast<void*>(context));
+
+  auto value = kv_store->read(Data::Utility::copyToString(key));
+  release_envoy_data(key);
+
+  return value.has_value() ? Data::Utility::copyToBridgeData(value.value()) : envoy_nodata;
+}
+
+void c_kv_store_save(envoy_data key, envoy_data value, const void* context) {
+  auto kv_store = *static_cast<KeyValueStoreSharedPtr*>(const_cast<void*>(context));
+
+  kv_store->save(Data::Utility::copyToString(key), Data::Utility::copyToString(value));
+  release_envoy_data(key);
+  release_envoy_data(value);
+}
+
+void c_kv_store_remove(envoy_data key, const void* context) {
+  auto kv_store = *static_cast<KeyValueStoreSharedPtr*>(const_cast<void*>(context));
+
+  kv_store->remove(Data::Utility::copyToString(key));
+  release_envoy_data(key);
+}
+
+} // namespace
+
+envoy_kv_store KeyValueStore::asEnvoyKeyValueStore() {
+  return envoy_kv_store{
+      &c_kv_store_read,
+      &c_kv_store_save,
+      &c_kv_store_remove,
+      new KeyValueStoreSharedPtr(shared_from_this()),
```

Comment:
So I needed to read the documentation of `enable_shared_from_this` to understand better what's happening here.  The documentation that I found https://en.cppreference.com/w/cpp/memory/enable_shared_from_this.

I am probably missing something but here is my question:
The documentation says: `It is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by std::shared_ptr<T>.`

When I look at the code from this PR I can see that we register a new KeyValueStore in [here](https://github.com/envoyproxy/envoy-mobile/pull/2430/files#diff-204f30e7778ed8d1f3e6008076942d0cd78f18df61826d44ef7f2d54b2b9d2fbR216) and that this code calls into [here](https://github.com/envoyproxy/envoy-mobile/pull/2430/files#diff-cbbc688841d6fdf3e2e1de0f4b000970b68527576c8928c653205fc2149cfe72R41) but I do not think that at this point our object is managed by any `shared_ptr` which documentation says is required for `shared_from_this` call to be safe. What am I missing here?

Reply:
Discussed offline - the `KeyValueStorer` managed with `shared_ptr` is coming from a user of the library and it's passed in [here](https://github.com/envoyproxy/envoy-mobile/pull/2430/files#diff-204f30e7778ed8d1f3e6008076942d0cd78f18df61826d44ef7f2d54b2b9d2fbR87)